### PR TITLE
👌 Improve notebook integration

### LINKED ISSUE #687: 👌 Better Notebook integration
* glotaran version: all

## Automatic rendering of markdown

Right now at the beginning of each notebook I have to define helper functions to render markdown.

https://github.com/glotaran/pyglotaran/blob/1de783ffaba9c3e45cfb162a01f3f9b000afe05f/docs/source/notebooks/quickstart/quickstart.ipynb#L19-L31

This can be prevented by adding the [special method `_repr_markdown_`](https://ipython.readthedocs.io/en/latest/config/integrating.html?highlight=_repr_markdown_#rich-display) so just having `model` (or any other object that supports markdown) in the last line of your cell renders it properly as markdown.

## Adjustable heading level for model

`jupyterlab>=3` has a builtin TOC (Table Of Content) which allows you to quickly navigate your notebook.
Since the heading level of `Model` in the markdown representation of model starts with `h1`, this messes up the TOC.
grafik

Wished output (model starts as `h3`)
grafik

We can easily solve this problem by adding a `base_heading_level` parameter to the `markdown` method.

## No markdown in `__repr__`
Not directly a notebook issue, but has to do with markdown.
Overwriting the magic method `__repr__` to use markdown leads to not helpful and cluttered test assertions/tracebacks.

Consider the following fake test file (`dummy_test.py`):
```python
from glotaran.parameter import ParameterGroup


def test_dummy():
    foo = ParameterGroup.from_dict({"foo": [1, 2, 3]})

    foo.get("bar")
```

Running `pytest` will result in this output:
```console
$ pytest dummy_test.py
================================================== test session starts ================================================== 
platform win32 -- Python 3.8.6, pytest-6.2.2, py-1.9.0, pluggy-0.13.1
benchmark: 3.2.3 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibrati

on_precision=10 warmup=False warmup_iterations=100000)
rootdir: D:\git\pyglotaran, configfile: tox.ini
plugins: anyio-2.1.0, nbval-0.9.6, benchmark-3.2.3, cov-2.11.1, env-0.6.2
collected 1 item

dummy_test.py F                                                                                                    [100%]

======================================================= FAILURES ======================================================== ______________________________________________________ test_dummy _______________________________________________________

self =   * __foo__:

    |   _Label_ |   _Value_ |   _StdErr_ |   _Min_ |   _Max_ | _Vary_   | _Non-Negative_   | _Expr_   |
... | None     |
    |         3 |         3 |          0 |    -inf |     inf | True     | False            | None     |


label = 'bar'

    def get(self, label: str) -> Parameter:
        """Gets a :class:`Parameter` by its label.

        Parameters
        ----------
        label :
            The label of the parameter, with its path in a parameter group prepended.
        """

        # sometimes the spec parser delivers the labels as int
        label = str(label)

        path = label.split(".")
        label = path.pop()

        # TODO: audit this code
        group = self
        for element in path:
            try:
                group = group[element]
            except KeyError:
                raise ParameterNotFoundException(path, label)
        try:
>           return group._parameters[label]
E           KeyError: 'bar'

..\glotaran\parameter\parameter_group.py:293: KeyError

During handling of the above exception, another exception occurred:

    def test_dummy():
        foo = ParameterGroup.from_dict({"foo": [1, 2, 3]})

>       foo.get("bar")

dummy_test.py:7:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self =   * __foo__:

    |   _Label_ |   _Value_ |   _StdErr_ |   _Min_ |   _Max_ | _Vary_   | _Non-Negative_   | _Expr_   |
... | None     |
    |         3 |         3 |          0 |    -inf |     inf | True     | False            | None     |


label = 'bar'

    def get(self, label: str) -> Parameter:
        """Gets a :class:`Parameter` by its label.

        Parameters
        ----------
        label :
            The label of the parameter, with its path in a parameter group prepended.
        """

        # sometimes the spec parser delivers the labels as int
        label = str(label)

        path = label.split(".")
        label = path.pop()

        # TODO: audit this code
        group = self
        for element in path:
            try:
                group = group[element]
            except KeyError:
                raise ParameterNotFoundException(path, label)
        try:
            return group._parameters[label]
        except KeyError:
>           raise ParameterNotFoundException(path, label)
E           glotaran.parameter.parameter_group.ParameterNotFoundException: Cannot find parameter .bar

..\glotaran\parameter\parameter_group.py:295: ParameterNotFoundException
=================================================== warnings summary ==================================================== c:\anaconda3\envs\pyglotaran\lib\site-packages\win32\lib\pywintypes.py:2
  c:\anaconda3\envs\pyglotaran\lib\site-packages\win32\lib\pywintypes.py:2: DeprecationWarning: the imp module is deprecat

ed in favour of importlib; see the module's documentation for alternative uses
    import imp, sys, os

-- Docs: https://docs.pytest.org/en/stable/warnings.html
================================================ short test summary info ================================================ FAILED dummy_test.py::test_dummy - glotaran.parameter.parameter_group.ParameterNotFoundException: Cannot find parameter...


============================================= 1 failed, 1 warning in 2.64s ==============================================
```

### PR DESCRIPTION
**Pull request by s-weigand**
This PR makes jupyter notebooks first-class citizens.
- `Model.markdown` now has a configurable `base_heading_level` not to break the notebook TOC
- `Model`, `ParameterGroup`, `Result`, `Scheme` and `KMatrix` render automatically as markdown
- `markdown` methods render automatically
- Added convenience function `display_file` allows showing files with syntax highlighting
- `ParameterGroup` and `Parameter` have a `__repr__` that allows constructing the object

**Testing**

Passing the tests is mandatory.

**Closing issues**

closes #687

### CODE REVIEW

File: glotaran/utils/ipython.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+"""Glotaran module with utilities for ``ipython`` integration (e.g. ``notebooks``)."""
+from __future__ import annotations
+
+from collections import UserString
+from os import PathLike
+
+
+class MarkdowStr(UserString):
```
Comment (jsnel) [inline glotaran/utils/ipython.py:8]: MarkDow*n*Str

File: glotaran/model/base_model.py [hunk 2]
```diff
@@ -6,6 +6,7 @@
 
 from glotaran.deprecation import deprecate
 from glotaran.parameter import ParameterGroup
+from glotaran.utils.ipython import MarkdowStr
```
Comment (jsnel) [inline glotaran/model/base_model.py:9]: Markdow*n*Str

File: glotaran/builtin/models/kinetic_image/k_matrix.py [hunk 3]
```diff
@@ -11,6 +11,7 @@
 from glotaran.builtin.models.kinetic_image.initial_concentration import InitialConcentration
 from glotaran.model import model_attribute
 from glotaran.parameter import Parameter
+from glotaran.utils.ipython import MarkdowStr
```
Comment (jsnel) [inline glotaran/builtin/models/kinetic_image/k_matrix.py:14]: Markdow*n*Str

File: glotaran/builtin/models/kinetic_image/k_matrix.py [hunk 4]
```diff
@@ -76,7 +77,7 @@ def matrix_as_markdown(
         self,
         compartments: list[str] = None,
         fill_parameters: bool = False,
-    ) -> str:
+    ) -> MarkdowStr:
```
Comment (jsnel) [inline glotaran/builtin/models/kinetic_image/k_matrix.py:80]: Markdow*n*Str

File: glotaran/builtin/models/kinetic_image/k_matrix.py [hunk 5]
```diff
@@ -106,7 +107,11 @@ def matrix_as_markdown(
             )
         return self._array_as_markdown(array, compartments, compartments)
 
-    def a_matrix_as_markdown(self, initial_concentration: InitialConcentration) -> str:
+    def _repr_markdown_(self) -> str:
+        """Special method used by ``ipython`` to render markdown."""
+        return str(self.matrix_as_markdown())
+
+    def a_matrix_as_markdown(self, initial_concentration: InitialConcentration) -> MarkdowStr:
```
Comment (jsnel) [inline glotaran/builtin/models/kinetic_image/k_matrix.py:114]: Markdow*n*Str

File: glotaran/builtin/models/kinetic_image/k_matrix.py [hunk 6]
```diff
@@ -124,7 +129,7 @@ def a_matrix_as_markdown(self, initial_concentration: InitialConcentration) -> s
         )
 
     @staticmethod
-    def _array_as_markdown(array, row_header, column_header):
+    def _array_as_markdown(array, row_header, column_header) -> MarkdowStr:
```
Comment (jsnel) [inline glotaran/builtin/models/kinetic_image/k_matrix.py:132]: Markdow*n*Str

File: glotaran/builtin/models/kinetic_image/k_matrix.py [hunk 7]
```diff
@@ -142,7 +147,7 @@ def _array_as_markdown(array, row_header, column_header):
 
             markdown += "|\n"
 
-        return markdown
+        return MarkdowStr(markdown)
```
Comment (jsnel) [inline glotaran/builtin/models/kinetic_image/k_matrix.py:150]: Markdow*n*Str

File: glotaran/builtin/models/kinetic_image/k_matrix.py (hunk 3 above)
Comment (s-weigand) [reply glotaran/builtin/models/kinetic_image/k_matrix.py]: Fixed during rebase

File: glotaran/utils/ipython.py (hunk 1 above)
Comment (s-weigand) [reply glotaran/utils/ipython.py]: Fixed during rebase

**Comment by s-weigand** [comment]
Since I rebased on `main` anyway (which means losing the reference points to your comments) I also clean up the typos in an interactive rebase, as if it did never happen 😝

**Review by jsnel — APPROVED** [review]
LGTM!

**Pull request merged**